amd_openvx_extensions/amd_rpp/source/Copy.cpp

```diff
@@ -0,0 +1,233 @@
+    STATUS_ERROR_CHECK(vxSetMetaFormatAttribute(metas[1], VX_TENSOR_NUMBER_OF_DIMS, &num_tensor_dims, sizeof(num_tensor_dims)));
+    STATUS_ERROR_CHECK(vxSetMetaFormatAttribute(metas[1], VX_TENSOR_DIMS, &tensor_dims, sizeof(tensor_dims)));
+    STATUS_ERROR_CHECK(vxSetMetaFormatAttribute(metas[1], VX_TENSOR_DATA_TYPE, &tensor_type, sizeof(tensor_type)));
+    STATUS_ERROR_CHECK(vxSetMetaFormatAttribute(metas[1], VX_TENSOR_FIXED_POINT_POSITION, &tensor_fixed_point_position, sizeof(tensor_fixed_point_position)));
+    vxReleaseTensor(&output);
+    vxReleaseParameter(&output_param);
+    return status;
+}
+
+static vx_status VX_CALLBACK processCopy(vx_node node, const vx_reference *parameters, vx_uint32 num)
+{
+    vx_status status = VX_SUCCESS;
+    CopyLocalData *data = NULL;
+    STATUS_ERROR_CHECK(vxQueryNode(node, VX_NODE_LOCAL_DATA_PTR, &data, sizeof(data)));
+
+    if (data->device_type == AGO_TARGET_AFFINITY_GPU)
+    {
+#if ENABLE_OPENCL
+        cl_command_queue handle = data->handle.cmdq;
+        refreshCopy(node, parameters, num, data);
+        clEnqueueCopyBuffer(handle, data->cl_pSrc, data->cl_pDst, 0, 0, data->tensor_size, 0, NULL, NULL);
+#elif ENABLE_HIP
+        refreshCopy(node, parameters, num, data);
+        hipMemcpy(data->hip_pDst, data->hip_pSrc, data->tensor_size, hipMemcpyDeviceToDevice);
+#endif
+    }
+    if (data->device_type == AGO_TARGET_AFFINITY_CPU)
+    {
+        refreshCopy(node, parameters, num, data);
+        memcpy(data->pDst, data->pSrc, data->tensor_size);
+    }
+    return status;
+}
+
+static vx_status VX_CALLBACK initializeCopy(vx_node node, const vx_reference *parameters, vx_uint32 num)
+{
+    CopyLocalData *data = new CopyLocalData;
+    memset(data, 0, sizeof(*data));
+#if ENABLE_OPENCL
+    STATUS_ERROR_CHECK(vxQueryNode(node, VX_NODE_ATTRIBUTE_AMD_OPENCL_COMMAND_QUEUE, &data->handle.cmdq, sizeof(data->handle.cmdq)));
+#elif ENABLE_HIP
+    STATUS_ERROR_CHECK(vxQueryNode(node, VX_NODE_ATTRIBUTE_AMD_HIP_STREAM, &data->handle.hipstream, sizeof(data->handle.hipstream)));
+#endif
+    STATUS_ERROR_CHECK(vxCopyScalar((vx_scalar)parameters[2], &data->device_type, VX_READ_ONLY, VX_MEMORY_TYPE_HOST));
+    vx_size num_of_dims;
+    size_t tensor_dims[NUM_OF_DIMS];
+    STATUS_ERROR_CHECK(vxQueryTensor((vx_tensor)parameters[0], VX_TENSOR_NUMBER_OF_DIMS, &num_of_dims, sizeof(vx_size)));
+    STATUS_ERROR_CHECK(vxQueryTensor((vx_tensor)parameters[0], VX_TENSOR_DIMS, tensor_dims, sizeof(vx_size) * num_of_dims));
+    STATUS_ERROR_CHECK(vxQueryTensor((vx_tensor)parameters[0], VX_TENSOR_DATA_TYPE, &data->in_tensor_type, sizeof(data->in_tensor_type)));
+    STATUS_ERROR_CHECK(vxQueryTensor((vx_tensor)parameters[1], VX_TENSOR_DATA_TYPE, &data->out_tensor_type, sizeof(data->out_tensor_type)));
+    
+    data->tensor_size = 1;
+    for(int i = 0; i < num_of_dims; i++)
+        data->tensor_size *= tensor_dims[i];
```

Comment:
Looks like it is assuming strides to be same as dims which may not be the case

Reply:
In this case the Tensor size is calculated as N * C * H * W which is specified in the dimensions of the tensor. So that we can copy the entire batch of images at once.

Reply:
TODO: this won't be true of strides are different than dims